velox/exec/tests/SimpleAggregateAdapterTest.cpp

```diff
@@ -471,5 +483,324 @@ TEST_F(SimpleCountNullsAggregationTest, basic) {
   testAggregations({vectors}, {}, {"simple_count_nulls(c2)"}, {expected});
 }
 
+// A testing aggregation function that uses the function state.
```

Comment:
Hi @liujiayi771, thank you for iterating on the unit test! I find the unit test still a bit hard to understand. Could we simplify it a bit by separating the checks for different purposes?

We need to test the function-level states from three aspects:
1. function-level states work correctly with the simple UDAF interface.
2. companion functions receive correct function-level states and work properly.
3. the UDAF with function-level states work correctly in window operations.

So I propose that we construct the unit tests as follows:
1. We don't need to check the function state when processing every row. So we can just move the checks into the UDAF's initialize() method.
2. We can make the FunctionStateTestAggregate class templated by a `bool testCompanion`. Inside initialize(), if testCompanion is true, we check that rawInputType is always {bigint, bigint} when step is partial, and it's always {row<bigint, double>} when step is intermediate. Plus, the constantInput is set correctly when step is partial. On the other hand, if testCompanion is false, we check that rawInputType is always {bigint, bigint} and constantInput is set correctly for all steps.
3. Let the addInput() method not use the argument `increment` but the constant value in the function state instead. 
4. We can register a UDAF simple_function_state_agg_main with testCompanion being false. With this UDAF, we can use the testAggregations API to check that results are correct. This test would tell us that function-level state works properly with simple UDAFs. (You don't need testIncrementalAggregation and testStreaming since those focus more on individual accumulator design. You can turn them off by calling disableTestStreaming() and disableTestIncremental() so that you don't need to modify AggregationTestBase.cpp.)
5. We can register another UDAF simple_function_state_agg_companion with testCompanion being true. With this UDAF, we use testAggregationsWithCompanion() to check the results are correct.
6. Let's also use the testWindowFunction() API to test that the main UDAF works correctly in window operations.

What do you think?

Reply:
@kagamiori Thank you very much for providing the detailed suggestion. I will refactor the current test cases based on these suggestions.

Reply:
@kagamiori If we use `testAggregationsWithCompanion` to test companion aggregate functions, we will encounter a scenario where `simple_function_state_agg_companion_partial`'s actual step is `kFinal`. In this case, the `initialize` method passes in step `kPartial`. So I chose to use single aggregate to test companion aggregate functions before.

Should we add a `companionStep` to differentiate the actual step in function state?

Reply:
Hi @liujiayi771, that's good finding. Let's pass an additional companionStep to the initialize() method to make it clear what that aggregation step is doing.